File: src/http.rs
use magnus::{Error, Module, RModule, Ruby, TryConvert, Value, method};

// Defines constant registration, `into_ffi`/`from_ffi`, and handlers for
// Ruby's `to_s`, `==`, `eql?`, and `hash` methods.
// `wreq::Version` only exposes protocol strings through `Debug`, so static
// labels avoid allocating an intermediate `String`.
define_ruby_enum!(
    /// An HTTP version.
    Version,
    "Wreq::Version",
    wreq::Version,
    strings:
    HTTP_09 => "HTTP/0.9",
    HTTP_10 => "HTTP/1.0",
    HTTP_11 => "HTTP/1.1",
    HTTP_2 => "HTTP/2.0",
    HTTP_3 => "HTTP/3.0",
);

// Defines constant registration, `into_ffi`/`from_ffi`, and handlers for
// Ruby's `to_s`, `to_sym`, `==`, `eql?`, and `hash` methods.
define_ruby_enum!(
    /// An HTTP method.
    Method,
    "Wreq::Method",
    wreq::Method,
    symbols:
    GET => "get",
    HEAD => "head",
    POST => "post",
    PUT => "put",
    DELETE => "delete",
    OPTIONS => "options",
    TRACE => "trace",
    PATCH => "patch",
);

/// HTTP status code.
#[derive(Clone, Copy, PartialEq, Eq, Hash)]
#[magnus::wrap(class = "Wreq::StatusCode", free_immediately, size)]
pub struct StatusCode(pub wreq::StatusCode);

impl TryConvert for Version {
    fn try_convert(value: magnus::Value) -> Result<Self, magnus::Error> {
        <&Version>::try_convert(value).cloned()
    }
}

// ===== impl StatusCode =====

impl StatusCode {
    /// Return the status code as an integer.
    #[inline]
    pub const fn as_int(&self) -> u16 {
        self.0.as_u16()
    }

    /// Check if status is within 100-199.
    #[inline]
    pub fn is_informational(&self) -> bool {
        self.0.is_informational()
    }

    /// Check if status is within 200-299.
    #[inline]
    pub fn is_success(&self) -> bool {
        self.0.is_success()
    }

    /// Check if status is within 300-399.
    #[inline]
    pub fn is_redirection(&self) -> bool {
        self.0.is_redirection()
    }

    /// Check if status is within 400-499.
    #[inline]
    pub fn is_client_error(&self) -> bool {
        self.0.is_client_error()
    }

    /// Check if status is within 500-599.
    #[inline]
    pub fn is_server_error(&self) -> bool {
        self.0.is_server_error()
    }

    /// Convert status code to string.
    #[inline]
    pub fn to_s(&self) -> String {
        self.0.to_string()
    }

    /// Value-based equality for Ruby (`==`).
    #[inline]
    pub fn equals(&self, other: Value) -> bool {
        <&StatusCode>::try_convert(other)
            .map(|other| *self == *other)
            .unwrap_or(false)
    }

    /// Strict equality for Ruby (`eql?`).
    #[inline]
    pub fn is_eql(&self, other: Value) -> bool {
        self.equals(other)
    }

    /// Hash value for Ruby (`hash`).
    #[inline]
    pub fn hash_value(&self) -> u64 {
        use std::hash::{Hash, Hasher};
        let mut hasher = std::collections::hash_map::DefaultHasher::new();
        self.0.hash(&mut hasher);
        hasher.finish()
    }

    /// Return the status code as an integer (Ruby `to_i`).
    #[inline]
    pub const fn to_i(&self) -> u16 {
        self.0.as_u16()
    }
}

impl From<wreq::StatusCode> for StatusCode {
    fn from(status: wreq::StatusCode) -> Self {
        Self(status)
    }
}

pub fn include(ruby: &Ruby, gem_module: &RModule) -> Result<(), Error> {
    let method_class = gem_module.define_class("Method", ruby.class_object())?;
    Method::define_constants(method_class)?;
    method_class.define_method("to_s", method!(Method::to_s, 0))?;
    method_class.define_method("to_sym", method!(Method::to_sym, 0))?;
    method_class.define_method("==", method!(Method::equals, 1))?;
    method_class.define_method("eql?", method!(Method::is_eql, 1))?;
    method_class.define_method("hash", method!(Method::hash_value, 0))?;

    let version_class = gem_module.define_class("Version", ruby.class_object())?;
    Version::define_constants(version_class)?;
    version_class.define_method("to_s", method!(Version::to_s, 0))?;
    version_class.define_method("==", method!(Version::equals, 1))?;
    version_class.define_method("eql?", method!(Version::is_eql, 1))?;
    version_class.define_method("hash", method!(Version::hash_value, 0))?;

    let status_code_class = gem_module.define_class("StatusCode", ruby.class_object())?;
    status_code_class.define_method("as_int", method!(StatusCode::as_int, 0))?;
    status_code_class.define_method("informational?", method!(StatusCode::is_informational, 0))?;
    status_code_class.define_method("success?", method!(StatusCode::is_success, 0))?;
    status_code_class.define_method("redirection?", method!(StatusCode::is_redirection, 0))?;
    status_code_class.define_method("client_error?", method!(StatusCode::is_client_error, 0))?;
    status_code_class.define_method("server_error?", method!(StatusCode::is_server_error, 0))?;
    status_code_class.define_method("to_s", method!(StatusCode::to_s, 0))?;
    status_code_class.define_method("==", method!(StatusCode::equals, 1))?;
    status_code_class.define_method("eql?", method!(StatusCode::is_eql, 1))?;
    status_code_class.define_method("hash", method!(StatusCode::hash_value, 0))?;
    status_code_class.define_method("to_i", method!(StatusCode::to_i, 0))?;

    Ok(())
}

File: lib/wreq_ruby/http.rb
# frozen_string_literal: true

module Wreq
  # HTTP method enumeration backed by Rust.
  #
  # Variants are exposed as constants under this class.
  # Each constant is an instance of {Wreq::Method}.
  #
  # @example Using predefined constants
  #   method = Wreq::Method::GET
  #   method.class #=> Wreq::Method
  #
  # @example In request context
  #   Wreq.request(url: "https://api.example.com", method: Wreq::Method::POST)
  class Method
    # Constants are set by the native extension at initialization.
    # These stubs are for documentation only.
    unless const_defined?(:GET)
      GET = nil # @return [Wreq::Method] HTTP GET method
      HEAD = nil # @return [Wreq::Method] HTTP HEAD method
      POST = nil # @return [Wreq::Method] HTTP POST method
      PUT = nil # @return [Wreq::Method] HTTP PUT method
      DELETE = nil # @return [Wreq::Method] HTTP DELETE method
      OPTIONS = nil # @return [Wreq::Method] HTTP OPTIONS method
      TRACE = nil # @return [Wreq::Method] HTTP TRACE method
      PATCH = nil # @return [Wreq::Method] HTTP PATCH method
    end

    # Returns the HTTP method token (e.g. "GET", "POST").
    # @return [String]
    unless method_defined?(:to_s)
      def to_s
      end
    end

    # Returns the HTTP method as a lowercase symbol (e.g. :get, :post).
    # @return [Symbol]
    unless method_defined?(:to_sym)
      def to_sym
      end
    end

    # Value-based equality. Returns true when both represent the same HTTP method.
    # @param other [Object]
    # @return [Boolean]    
    unless method_defined?(:==)
      def ==(other)
      end
    end

    # Strict equality for Hash key and Set member semantics.
    # @param other [Object]
    # @return [Boolean]
    unless method_defined?(:eql?)
      def eql?(other)
      end
    end

    # Hash value consistent with {#eql?} for use as Hash keys.
    # @return [Integer]
    unless method_defined?(:hash)
      def hash
      end
    end
  end

  # HTTP version enumeration backed by Rust.
  #
  # @example Using predefined constants
  #   version = Wreq::Version::HTTP_11
  #   version.class #=> Wreq::Version
  class Version
    # Constants are set by the native extension at initialization.
    # These stubs are for documentation only.
    unless const_defined?(:HTTP_11)
      HTTP_09 = nil # @return [Wreq::Version] HTTP/0.9
      HTTP_10 = nil # @return [Wreq::Version] HTTP/1.0
      HTTP_11 = nil # @return [Wreq::Version] HTTP/1.1
      HTTP_2 = nil # @return [Wreq::Version] HTTP/2
      HTTP_3 = nil # @return [Wreq::Version] HTTP/3
    end

    # Returns a string representation of the HTTP version.
    # @return [String] HTTP version as string
    unless method_defined?(:to_s)
      def to_s
      end
    end

    # Compares HTTP versions by semantic value, not object identity.
    #
    # This method is implemented by the native extension.
    # When comparing with non-{Wreq::Version} objects, it returns false.
    #
    # @param other [Object] object to compare against
    # @return [Boolean] true when both represent the same HTTP version
    # @example
    #   Wreq::Version::HTTP_11 == response.version
    unless method_defined?(:==)
      def ==(other)
      end
    end

    # Strict equality for Hash key and Set member semantics.
    # @param other [Object]
    # @return [Boolean]
    unless method_defined?(:eql?)
      def eql?(other)
      end
    end

    # Hash value consistent with {#eql?} for use as Hash keys.
    # @return [Integer]
    unless method_defined?(:hash)
      def hash
      end
    end
  end

  # HTTP status code wrapper.
  #
  # This class wraps standard HTTP status codes and provides
  # convenient methods to check the response category.
  #
  # The actual implementation is provided by Rust for performance.
  #
  # @example Check if response is successful
  #   status = response.status
  #   if status.success?
  #     puts "Request succeeded with code: #{status.as_int}"
  #   end
  #
  # @example Check different status categories
  #   status.informational?  # 1xx
  #   status.success?        # 2xx
  #   status.redirection?    # 3xx
  #   status.client_error?   # 4xx
  #   status.server_error?   # 5xx
  unless const_defined?(:StatusCode)
    class StatusCode
      # Returns the status code as an integer.
      #
      # @return [Integer] the numeric HTTP status code (100-599)
      def as_int
      end

      # Checks if status code is informational (1xx).
      #
      # Informational responses indicate that the request was received
      # and the process is continuing.
      #
      # @return [Boolean] true if status is 100-199
      def informational?
      end

      # Checks if status code indicates success (2xx).
      #
      # Success responses indicate that the request was successfully
      # received, understood, and accepted.
      #
      # @return [Boolean] true if status is 200-299
      def success?
      end

      # Checks if status code indicates redirection (3xx).
      #
      # Redirection responses indicate that further action needs to be
      # taken to complete the request.
      #
      # @return [Boolean] true if status is 300-399
      def redirection?
      end

      # Checks if status code indicates client error (4xx).
      #
      # Client error responses indicate that the request contains bad
      # syntax or cannot be fulfilled.
      #
      # @return [Boolean] true if status is 400-499
      def client_error?
      end

      # Checks if status code indicates server error (5xx).
      #
      # Server error responses indicate that the server failed to
      # fulfill a valid request.
      #
      # @return [Boolean] true if status is 500-599
      def server_error?
      end

      # Returns a string representation of the status code.
      # @return [String] Status code as string
      def to_s
      end

      # Returns the status code as an integer.
      # @return [Integer] the numeric HTTP status code
      def to_i
      end

      # Value-based equality. Only compares with other StatusCode instances.
      # @param other [Object]
      # @return [Boolean]
      def ==(other)
      end

      # Strict equality for Hash key and Set member semantics.
      # @param other [Object]
      # @return [Boolean]
      def eql?(other)
      end

      # Hash value consistent with {#eql?} for use as Hash keys.
      # @return [Integer]
      def hash
      end
    end
  end
end

module Wreq
  class StatusCode
    def inspect
      "#<Wreq::StatusCode #{self}>"
    end
  end

  class Version
    def inspect
      "#<Wreq::Version #{self}>"
    end
  end
end
